// eas-cli/build/commands/env/link.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const EasCommand_1 = tslib_1.__importDefault(require("../../commandUtils/EasCommand"));
const flags_1 = require("../../commandUtils/flags");
const EnvironmentVariableMutation_1 = require("../../graphql/mutations/EnvironmentVariableMutation");
const EnvironmentVariablesQuery_1 = require("../../graphql/queries/EnvironmentVariablesQuery");
const log_1 = tslib_1.__importDefault(require("../../log"));
const projectUtils_1 = require("../../project/projectUtils");
const prompts_1 = require("../../prompts");
const prompts_2 = require("../../utils/prompts");
class EnvironmentVariableLink extends EasCommand_1.default {
    async runAsync() {
        let { flags: { name, 'non-interactive': nonInteractive, environment }, } = await this.parse(_a);
        const { privateProjectConfig: { projectId }, loggedIn: { graphqlClient }, } = await this.getContextAsync(_a, {
            nonInteractive,
        });
        const projectDisplayName = await (0, projectUtils_1.getDisplayNameForProjectIdAsync)(graphqlClient, projectId);
        const variables = await EnvironmentVariablesQuery_1.EnvironmentVariablesQuery.sharedAsync(graphqlClient, {
            appId: projectId,
        });
        if (!name) {
            name = await (0, prompts_1.selectAsync)('Select shared variable', variables.map(variable => ({
                title: variable.name,
                value: variable.name,
            })));
        }
        const selectedVariable = variables.find(variable => variable.name === name);
        if (!selectedVariable) {
            throw new Error(`Shared variable ${name} doesn't exist`);
        }
        if (!environment) {
            environment = await (0, prompts_2.promptVariableEnvironmentAsync)(nonInteractive);
        }
        const linkedVariable = await EnvironmentVariableMutation_1.EnvironmentVariableMutation.linkSharedEnvironmentVariableAsync(graphqlClient, selectedVariable.id, projectId, environment);
        if (!linkedVariable) {
            throw new Error(`Could not link variable with name ${selectedVariable.name} to project with id ${projectId}`);
        }
        log_1.default.withTick(`Linked variable ${chalk_1.default.bold(linkedVariable.name)} to project ${chalk_1.default.bold(projectDisplayName)}.`);
    }
}
_a = EnvironmentVariableLink;
EnvironmentVariableLink.description = 'link a shared environment variable to the current project';
EnvironmentVariableLink.hidden = true;
EnvironmentVariableLink.flags = {
    name: core_1.Flags.string({
        description: 'Name of the variable',
    }),
    ...flags_1.EASEnvironmentFlag,
    ...flags_1.EASNonInteractiveFlag,
};
EnvironmentVariableLink.contextDefinition = {
    ..._a.ContextOptions.ProjectConfig,
    ..._a.ContextOptions.LoggedIn,
};
exports.default = EnvironmentVariableLink;
