// @expo/eas-build-job/dist/generic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Generic = void 0;
const zod_1 = require("zod");
const logger_1 = require("@expo/logger");
const common_1 = require("./common");
const step_1 = require("./step");
var Generic;
(function (Generic) {
    const BuilderEnvironmentSchemaZ = zod_1.z.object({
        image: zod_1.z.string(),
        node: zod_1.z.string().optional(),
        yarn: zod_1.z.string().optional(),
        pnpm: zod_1.z.string().optional(),
        bun: zod_1.z.string().optional(),
        env: zod_1.z.record(zod_1.z.string()),
        // Linux
        ndk: zod_1.z.string().optional(),
        // macOS
        bundler: zod_1.z.string().optional(),
        fastlane: zod_1.z.string().optional(),
        cocoapods: zod_1.z.string().optional(),
    });
    const CommonJobZ = zod_1.z.object({
        projectArchive: common_1.ArchiveSourceSchemaZ,
        secrets: zod_1.z.object({
            robotAccessToken: zod_1.z.string(),
            environmentSecrets: zod_1.z.array(common_1.EnvironmentSecretZ),
        }),
        expoDevUrl: zod_1.z.string().url(),
        builderEnvironment: BuilderEnvironmentSchemaZ,
        // We use this to discern between Android.Job, Ios.Job and Generic.Job.
        platform: zod_1.z.never().optional(),
        type: zod_1.z.never().optional(),
        triggeredBy: zod_1.z.literal(common_1.BuildTrigger.GIT_BASED_INTEGRATION),
        loggerLevel: zod_1.z.nativeEnum(logger_1.LoggerLevel).optional(),
    });
    const PathJobZ = CommonJobZ.extend({
        customBuildConfig: zod_1.z.object({
            path: zod_1.z.string(),
        }),
        steps: zod_1.z.never().optional(),
    });
    const StepsJobZ = CommonJobZ.extend({
        customBuildConfig: zod_1.z.never().optional(),
        steps: zod_1.z.array(step_1.StepZ).min(1),
    });
    Generic.JobZ = zod_1.z.union([PathJobZ, StepsJobZ]);
    Generic.PartialJobZ = zod_1.z.union([PathJobZ.partial(), StepsJobZ.partial()]);
})(Generic || (exports.Generic = Generic = {}));
